### PR TITLE
Jz dev

### PR DESCRIPTION
**Pull request by Howon**

### CODE REVIEW

File: routes_parse.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+               seq_in > json_new[ref]["seq_out"]):
+                json_new[ref] = {}
+                json_new[ref]["color"] = data["color"]
+                json_new[ref]["points"] = data["points"]
+                json_new[ref]["seq_out"] = seq_in
+                # for pt in data["points"]:
+                #     if pt not in json_new[ref]["points"]:
+                #         json_new[ref]["points"].append(pt)
+        geojson_new = {"type": "FeatureCollection", "features": []}
+        for ref in json_new:
```
Comment (alanhdu) [inline routes_parse.py]: ...why do we need two for loops? Why not directly construct `geojson_new` in the first place?

File: static/js/livesubway.js [hunk 2]
```diff
@@ -155,6 +259,7 @@ const fetchMap = (fetcher, map, routes, finish) => {
 
   const routePromise = new Promise((resolve, reject) => {
     fetcher("/map_geojson", (stopData) => {
+
```
Comment (alanhdu) [inline static/js/livesubway.js]: ?

File: static/js/livesubway.js [hunk 3]
```diff
@@ -140,12 +243,13 @@ const getJSON = (path, success, fail) => {
 
 const fetchMap = (fetcher, map, routes, finish) => {
   const renderRoutes = (routesData, cb) => {
+
```
Comment (alanhdu) [inline static/js/livesubway.js]: ?

Comment (Howon) [inline static/js/livesubway.js]: sup

Comment (Howon) [inline static/js/livesubway.js]: fetcher was used before when our Geojson data was massive so
I was implementing IndexedDB front end cache. this was meant
to take that in as a generic retrieval function that bottom main
method passes in. Fetcher could fetch from either client-side cache
or by doing a call to the server

File: static/js/livesubway.js (hunk 2 above)
Comment (Howon) [inline static/js/livesubway.js]: same as above

Comment (Howon) [inline static/js/livesubway.js]: prob should do stopData =>  instead of (stopData) =>

File: routes_parse.py (hunk 1 above)
Comment (jonathanzhang99) [inline routes_parse.py]: First for loop needs to find the best collection of points to use and the second constructs the actual json object.

File: app.py [hunk 4]
```diff
@@ -118,18 +119,25 @@ def subway_cars():
...
-        demo_emit = demos[counter % len(demos)]
-        print demo_emit
-        print "Emitted."
-        socketio.emit('feed', demo_emit)
-        counter += 1
+        # demo_emit = demos[counter % len(demos)]
+        # print demo_emit
+        # print "Emitted."
+        # socketio.emit('feed', demo_emit)
+        # counter += 1
```
Comment (Howon) [inline app.py]: It's his first time using socket.io so I will let this pass

File: static/js/livesubway.js [hunk 5]
```diff
@@ -93,6 +93,12 @@ const findDistance = (startindex, endindex, coordmap) => {
 // Calculates the speed of the train required to get 
 // from <startindex:int> to <endindex:int> along the specific line
 // within SERVER_DELAY seconds.
+/**
+ * @param  {[type]}
+ * @param  {[type]}
+ * @param  {[type]}
+ * @return {[type]}
```
Comment (alanhdu) [inline static/js/livesubway.js]: ?

File: parse_schedule.py [hunk 6]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+    with open(MAP_DIR + "times.json", "w") as times_fout:
+        json.dump(times_jout, times_fout)
 
 # TODO: Add argument parser
 with open(TRANSIT_DIR + "trips.txt", "r") as trips_fin, \
         open(TRANSIT_DIR + "stop_times.txt", "r") as times_fin, \
         open(MAP_DIR + "shapes.json", "r") as shapes_fin:
     parse_trips(trips_fin)
     parse_routes(shapes_fin)
+    parse_times(times_fin)
```
Comment (alanhdu) [inline parse_schedule.py]: Can we just separate this out into three different context managers? We should probably also use `os.path.join` instead of string concatenation for cross-platform compatibility: 

```python
with open(os.path.join(TRANSIT_DIR, "trips.txt")) as trips_fin:
    parse_trips(trips_fin)
with open(os.path.join(TRANSIT_DIR, "stop_times.txt")) as fin:
    parse_routes(fin)
...
```

File: app.py [hunk 7]
```diff
@@ -1,6 +1,8 @@
 import cPickle as pickle
 
+from datetime import datetime
 from eventlet import monkey_patch
+from eventlet.greenthread import spawn, sleep
```
Comment (alanhdu) [inline app.py]: If we're going to monkey patch stuff w/ eventlet, shouldn't `time.sleep` just work?

File: app.py [hunk 8]
```diff
@@ -23,14 +25,20 @@
...
-        open(JSON_DIR + "colors.json", "r") as colors_f:
+        open(JSON_DIR + "colors.json", "r") as colors_f, \
+        open(JSON_DIR + "times.json", "r") as times_f:
     graph = pickle.load(graph_f)
     prev_stops = pickle.load(prev_stops_f)
     shapes = json.load(shapes_f)
     stops = json.load(stops_f)
     routes = json.load(routes_f)
     colors = json.load(colors_f)
+    times = json.load(times_f)
```
Comment (alanhdu) [inline app.py]: Can we not nest all these context parameters and split them up?

File: app.py [hunk 9]
```diff
@@ -23,14 +25,20 @@
...
 
+
+def log(msg):
+    print "[{}] {}".format(str(datetime.now().replace(microsecond=0)),
+                           msg)
```
Comment (alanhdu) [inline app.py]: Can we use the builtin `logging` module instead?

File: app.py [hunk 10]
```diff
@@ -111,14 +114,25 @@ def stops_json():
...
 
+def schedule_timer():
+    while True:
+        spawn(schedule_handler)
+        sleep(5)
+
+
+def schedule_handler():
+    log("server emmited hello world")
+    socketio.emit("update", {"hello": "world"})
```
Comment (alanhdu) [inline app.py]: Why not just inline this function? Seems like premature abstraction to me.

File: parse_schedule.py [hunk 11]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+
+
+def parse_times(times_fin):
+    """Creates a JSON dictionary. This file is redundant with trips_fout
 
+    Args:
+        times_fin (TYPE): Description
+
+    Returns:
+        TYPE: Description
```
Comment (alanhdu) [inline parse_schedule.py]: ?

File: parse_schedule.py [hunk 12]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
-        json.dump(trips_jout, trips_fot)
+    with open(MAP_DIR + "/trips.json", "w") as trips_fout:
+        json.dump(trips_jout, trips_fout)
+
+
+def parse_times(times_fin):
+    """Creates a JSON dictionary. This file is redundant with trips_fout
 
+    Args:
+        times_fin (TYPE): Description
```
Comment (alanhdu) [inline parse_schedule.py]: Fill in type and description

File: parse_schedule.py [hunk 13]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+
+    Returns:
+        TYPE: Description
+    """
+    times_fin.readline()
+    times_jout = {"WKD": [], "SAT": [], "SUN": []}
```
Comment (alanhdu) [inline parse_schedule.py]: Better variable name needed.

In generally, I really dislike this the naming scheme here -- It's somehow cryptic and verbose at the same time (`sbw_info` is a good example -- cryptic (I assume `sbw` is for `subway`?) and unclear (`info` is such a generic name...)).

File: parse_schedule.py [hunk 14]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+def parse_times(times_fin):
+    """Creates a JSON dictionary. This file is redundant with trips_fout
 
+    Args:
+        times_fin (TYPE): Description
+
+    Returns:
+        TYPE: Description
+    """
+    times_fin.readline()
```
Comment (alanhdu) [inline parse_schedule.py]: Document why we're skipping the first line.

File: parse_schedule.py [hunk 15]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+    times_jout = {"WKD": [], "SAT": [], "SUN": []}
+    curr_trip_id = ""
+    curr_trip = []
+    for line in times_fin:
+        line = line.split(",")
+        if curr_trip_id == "":
```
Comment (alanhdu) [inline parse_schedule.py]: I would use `None` as the sentinel value instead.

File: parse_schedule.py [hunk 16]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+            curr_trip_id = line[0]
+        # check to see if the subway trip has changed - all trips are
+        # expected to be in order
+        if curr_trip_id != line[0]:
```
Comment (alanhdu) [inline parse_schedule.py]: `elif`?

File: parse_schedule.py [hunk 17]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+            sbw_info = curr_trip_id.split("_")
```
Comment (alanhdu) [inline parse_schedule.py]: Can we use a tuple destructuring assignment here instead?

File: parse_schedule.py [hunk 18]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+            sbw_day = sbw_info[0][-3:]
+            assert(sbw_day in ["SAT", "SUN", "WKD"])
```
Comment (alanhdu) [inline parse_schedule.py]: Should probably just be:
```python
assert sbw_day in times_jout
```

Comment (alanhdu) [inline parse_schedule.py]: Why do we need an `assert` in the first place? Wouldn't we get a `KeyNotFound` (or whatever the exception is) anyways?

File: parse_schedule.py [hunk 19]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+            times_jout[sbw_day].append({
+                "line": sbw_info[-1][0:sbw_info[-1].find('.')],
+                "id": curr_trip_id,
+                "init_time": convertTime(sbw_info[1]),
+                "trip_time": curr_trip
+            })
+            assert(times_jout[sbw_day][0]["init_time"] ==
+                   times_jout[sbw_day][0]["trip_time"][0][0])
```
Comment (alanhdu) [inline parse_schedule.py]: Add comment explaining this assert

File: parse_schedule.py [hunk 20]
```diff
@@ -97,13 +103,56 @@ def parse_trips(trips_fin):
...
+
+            curr_trip_id, curr_trip = line[0], []
+
+        curr_trip.append((line[1], line[3]))
+    for sbw_day, sbw_data in times_jout.iteritems():
+        times_jout[sbw_day] = sorted(sbw_data, key=lambda x:
+                                     (int(x.get("id").split('_')[1]),
+                                      x.get("line")))
```
Comment (alanhdu) [inline parse_schedule.py]: Why `get` instead of just `[]`? Are these keys not guaranteed to be in `x`?

Also, I prefer the `lambda` to be all in one line, not broken up over several.

File: app.py (hunk 10 above)
Comment (jonathanzhang99) [inline app.py]: the function is currently a placeholder - but trying to make this process asynchronous by spawning a greenthread

File: parse_schedule.py (hunk 20 above)
Comment (jonathanzhang99) [inline parse_schedule.py]: its hard to get it all in one line with pep8 80 char limit and spacing rules

File: parse_schedule.py (hunk 6 above)
Comment (jonathanzhang99) [inline parse_schedule.py]: will do! also i was thinking of adding an ArgumentParser later to make running from CL easier

File: app.py (hunk 8 above)
Comment (jonathanzhang99) [inline app.py]: ?

File: parse_schedule.py (hunk 13 above)
Comment (jonathanzhang99) [inline parse_schedule.py]: what would you recommend?

File: app.py (hunk 8 above)
Comment (Howon) [inline app.py]: what alan means i think is have all these different values be JSON elements
in on JSON file instead of having separate json files

File: parse_schedule.py (hunk 20 above)
Comment (Howon) [inline parse_schedule.py]: +1

i would actually make it into its own function as this lambda seems particularly ugly

File: app.py (hunk 7 above)
Comment (jonathanzhang99) [inline app.py]: yup - still figuring out asynchrony in python

File: app.py (hunk 8 above)
Comment (jonathanzhang99) [inline app.py]: isn't it easier to organize if we have them separately?

Comment (Howon) [inline app.py]: is it?

also you can remove a few of these json files now i think
(use routes instead of shapes, etc)

File: app.py (hunk 9 above)
Comment (Howon) [inline app.py]: log > print

File: app.py (hunk 8 above)
Comment (jonathanzhang99) [inline app.py]: easier at least for now, if formatting changes need to be made to one of the json files

yah I can get rid of several of these

File: app.py [hunk 21]
```diff
@@ -114,16 +115,28 @@ def stops_json():
     return jsonify(stops)
 
 
+# Could also implement using decorator - would be more functional
```
Comment (jonathanzhang99) [inline app.py]: ignore

File: app.py [hunk 22]
```diff
@@ -2,6 +2,7 @@
 
 from datetime import datetime, date
 from eventlet import monkey_patch
+from eventlet.greenthread import spawn, sleep
```
Comment (jonathanzhang99) [inline app.py]: tmp for testing

File: app.py (hunk 10 above)
Comment (alanhdu) [inline app.py]: So the problem with just randomly spawning tasks is it's hard to deal with back pressure. For example, if you have:

```Python
while True:
    spawn(func)
    sleep(1)
```

but `func` takes more than 1 second to finish, then you'll have an unboundedly increasing number of `func` threads running.

File: app.py [hunk 23]
```diff
@@ -114,16 +115,28 @@ def stops_json():
     return jsonify(stops)
 
 
+# Could also implement using decorator - would be more functional
+def parse_time(t_str):
```
Comment (alanhdu) [inline app.py]: 👍

**Comment by alanhdu** [comment]
So regarding naming:

- I don't really like the [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation) thing you've got going on here, especially for types (e.g. `t_str` or `d_today` or whatever). IMO, it just adds a lot of unnecessary clutter.
- I find a lot of the abbreviations unnecessarily cute. Yes, `lo`, `hi`, `curr`, and `sbw` are a little shorter than `low`, `high`, `current`, and `subway`, but 3 characters really does not matter. This complaint is biggest for `sbw` and `b_search` -- `lo`, `hi`, and `curr` are at least commonly used abbreviations.
- Variables and functions should be named more semantically -- what do things _mean_ and _represent_, not just what they happen to be in a particular function (I think `b_search_for_curr_time` is an example of a not-so-good name -- like, I understand that it's a binary search, but why is that in the name? Why are we searching for`curr_time`?

(Sorry that I'm being super nitpicky -- I understand that it gets old pretty quickly).

File: app.py [hunk 24]
```diff
@@ -114,14 +114,77 @@ def stops_json():
...
+        weekday = "WKD"
+
+    curr_schedule = times[weekday]
+
+    c = b_search_for_curr_time(curr_schedule, datetime.today())
+    print c
+    print curr_schedule[c]
+
+
+def b_search_for_curr_time(arr, t_target):
```
Comment (alanhdu) [inline app.py]: IMO the logic here is also a bit convoluted and hard to follow. It's a relatively complicated thing you're doing, but maybe something like the following would be better:

```python
import datetime as dt
def find_next_train(arr, target, epsilon=dt.timedelta(seconds=30)):
    # document that we want to return some time t such that:
    #   t <= target < t + epsilon
    low, high = 0, len(arr)
    while True:
        if target < arr[low] or arr[high] + time.epsilon < target:
            raise ValueError("{} not found".format(target))
        elif arr[low] <= target < arr[low] + epsilon:
            return low
        else:
            mid = (low + high) // 2
            if arr[mid] + epsilon < target:
                low = mid
            else:
                high = mid
```

Comment (jonathanzhang99) [inline app.py]: kind of redundant since its already been changed, but I've simplified the binary search to effectively 8 lines of code. Yah I agree binary search is not really needed but the array is generally around 7k and can be up to 10k elements.

File: app.py [hunk 25]
```diff
@@ -114,14 +114,77 @@ def stops_json():
...
+    Returns:
+        str: day of the week to key times dictionary
+        int: index of next starting train in times[weekday]
+        list: currently underway and unfinished trips
+    """
+    weekday_int = datetime.today().weekday()
+
+    active_cars = []
+
+    if weekday_int is 5:
```
Comment (jonathanzhang99) [inline app.py]: yup - dictionary lookup is much better

**Comment by jonathanzhang99** [comment]
Is there a convention or a guide you would recommend?

**Comment by alanhdu** [comment]
Not really 😜. I think a lot of it comes down to personal taste + experience. I think https://a-nickels-worth.blogspot.de/2016/04/a-guide-to-naming-variables.html and http://journal.stuffwithstuff.com/2016/06/16/long-names-are-long/ have fairly good overviews, although they're not exactly comprehensive.

File: .gitignore [hunk 26]
```diff
@@ -9,6 +9,7 @@ API_KEYS.py
 .DS_Store
 .idea
 .ipynb_checkpoints
+.jshintrc
```
Comment (alanhdu) [inline .gitignore]: Why are we `.gitignore`-ing this? I *strongly* believe that `.jshintrc` should be checked in to the repo so that everyone is using the same style lints.

File: app.py [hunk 27]
```diff
@@ -2,15 +2,14 @@
...
 
 from flask import Flask, json, jsonify, render_template
-from flask_socketio import SocketIO, emit
+from flask_socketio import SocketIO
 
 from API_KEYS import mapbox_key
 from static import Edge, PrevStops, Segment, Stop, StopGraph, StopID  # noqa: F401
 
-import feed
+# import feed
```
Comment (alanhdu) [inline app.py]: Stray comment

File: app.py [hunk 28]
```diff
@@ -38,84 +37,25 @@
...
-#             "remaining_time": 25
-#         },
-#         {
-#             "path": [[-73.992629, 40.730328], [-73.989951, 40.734673]],
-#             "progress": 0.3,
-#             "remaining_time": 15
-#         }
-#     ]
-# ]
+    print "[{}] {}".format(str(datetime.now().replace(microsecond=0)), msg)
```
Comment (alanhdu) [inline app.py]: Is there a reason we're not using the built-in `logging` module instead?

File: app.py [hunk 29]
```diff
@@ -126,7 +66,37 @@ def parse_time(t_str):
     )
 
 
-def schedule_init():
+def schedule_daemon(index, weekday):
```
Comment (alanhdu) [inline app.py]: What is `index` here? Could we get a more semantically meaningful name?

File: app.py [hunk 30]
```diff
@@ -126,7 +66,37 @@ def parse_time(t_str):
...
+    today = datetime.today()
+    next_departure = parse_time(times[weekday][index]["init_time"])
+    time_delta = next_departure - today
```
Comment (alanhdu) [inline app.py]: I would just name this `delta` or `diff`

File: app.py [hunk 31]
```diff
@@ -126,7 +66,37 @@ def parse_time(t_str):
...
+                next_time = parse_time(times[weekday][index]["init_time"])
+            next_departure = next_time
+            log(str(next_departure))
+            socketio.emit("schedule", departures)
+            log("Emitted {} departures".format(len(departures)))
+            # write loop to check for equivalent times
+
+            time_delta = next_departure - datetime.today()
+        else:
+            raise ValueError(time_delta)
```
Comment (alanhdu) [inline app.py]: Raising an error if the next train is one day away seems pretty fishy to me. Why one day? Why not 1 hour? Or 30 minutes? What happens if there's track repair work and a train isn't running over the wekeend?

File: app.py [hunk 32]
```diff
@@ -126,7 +66,37 @@ def parse_time(t_str):
...
+            log("sleeping for {} seconds...".format(time_delta.seconds))
+            socketio.sleep(time_delta.seconds)
+            next_time = parse_time(times[weekday][index]["init_time"])
+            while next_time == next_departure:
+                departures.append(times[weekday][index])
+                if (index < len(times[weekday])):
+                    index += 1
+                else:
+                    raise Exception
+                    # TODO: implement functionality to change dates
```
Comment (alanhdu) [inline app.py]: ??

Also, be more specific about the type of `Exception`

File: app.py (hunk 31 above)
Comment (alanhdu) [inline app.py]: Also, I'd prefer that we create our own `Exception` classes (e.g. `class TrainNotFound(ValueException)` or whatever) and raise those, so we can distinguish _deliberate_ exceptions and unexpected ones.

File: app.py [hunk 33]
```diff
@@ -126,7 +66,37 @@ def parse_time(t_str):
...
+                next_time = parse_time(times[weekday][index]["init_time"])
+            next_departure = next_time
+            log(str(next_departure))
+            socketio.emit("schedule", departures)
+            log("Emitted {} departures".format(len(departures)))
+            # write loop to check for equivalent times
```
Comment (alanhdu) [inline app.py]: ?

File: app.py (hunk 29 above)
Comment (alanhdu) [inline app.py]: Ok -- I'm **really** unclear on what this function is supposed to be doing. I think it's supposed to be a clock that keeps emitting the "next" departure, but it seems very convoluted for that (i.e. why are we iterating over `index`? What happens when the the day-of-the-week changes?).

File: app.py [hunk 34]
```diff
@@ -126,7 +66,37 @@ def parse_time(t_str):
...
+    time_delta = next_departure - today
+
+    while True:
+        if time_delta.days == 0:
+            departures = []
+            log("sleeping for {} seconds...".format(time_delta.seconds))
+            socketio.sleep(time_delta.seconds)
+            next_time = parse_time(times[weekday][index]["init_time"])
+            while next_time == next_departure:
+                departures.append(times[weekday][index])
```
Comment (alanhdu) [inline app.py]: Unbounded growth of a list is :no_good:. Why do we need to keep the old departures?

File: app.py [hunk 35]
```diff
@@ -210,26 +171,12 @@ def subway_cars():
...
 
 def subway_cars_timer():
-    while True:
-        socketio.sleep(30)
-        # demo_emit = demos[counter % len(demos)]
-        # print demo_emit
-        # print "Emitted."
-        # socketio.emit('feed', demo_emit)
-        # counter += 1
+    pass
```
Comment (alanhdu) [inline app.py]: Just delete the function then.

File: parse_schedule.py [hunk 36]
```diff
@@ -51,110 +53,105 @@ def parse_routes(shapes_fin):
         json.dump(geojson_out, shapes_fout)
 
 
-def convertTime(sbw_time):
-    """Converts sbw_time string into a datetime time object
+def _convertTime(train_time):
```
Comment (alanhdu) [inline parse_schedule.py]: Why does this have an `_` now?

File: parse_schedule.py [hunk 37]
```diff
@@ -51,110 +53,105 @@ def parse_routes(shapes_fin):
...
+    """Converts train_time string into a datetime time object
 
     Args:
-        sbw_time (str): created from sbw_id.
+        train_time (str): created from train_id.
 
     Returns:
-        str: time in standard isoformat
+        str: time in standard isoformat HH:MM:SS
```
Comment (alanhdu) [inline parse_schedule.py]: ...Doesn't this return a `dt.time` object?

File: parse_schedule.py [hunk 38]
```diff
@@ -51,110 +53,105 @@ def parse_routes(shapes_fin):
...
-# TODO: Add argument parser
-with open(TRANSIT_DIR + "trips.txt", "r") as trips_fin, \
-        open(TRANSIT_DIR + "stop_times.txt", "r") as times_fin, \
-        open(MAP_DIR + "shapes.json", "r") as shapes_fin:
-    parse_trips(trips_fin)
-    parse_routes(shapes_fin)
-    parse_times(times_fin)
+def get_parser():
+    parser = ArgumentParser(
+        description="A script to write scheduled subway data")
```
Comment (alanhdu) [inline parse_schedule.py]: Weird line break + indentation here. What's wrong with just one line?

File: parse_static.py [hunk 39]
```diff
@@ -1002,7 +1002,7 @@ def write_static_files(args):
     print "Loading static schedule information..."
     loader = transitfeed.Loader(STATIC_TRANSIT_DIR)
     schedule = loader.Load()
-    print "Done. Writing to file(s)..."
+    print "Done.\n\nWriting to file(s)..."
```
Comment (alanhdu) [inline parse_static.py]: Just use multiple print statements
